### ISSUE TITLE
JSpecify: an anonymous class loses the `@Nullable` on its supertype's type argument

### ISSUE THREAD

**Issue by vlsi** [uber/NullAway#1746]
### Version

NullAway 0.14.0, and current `master` (21f02dd). Error Prone 2.50.0, JDK 21.

Flags: `-XepOpt:NullAway:JSpecifyMode=true -XDaddTypeAnnotationsToSymbol=true`.

### Summary

An anonymous class that extends `Base<@Nullable String>` has its overrides checked against `Base<String>`: the `@Nullable` in the supertype's type argument is dropped. Every override whose signature wraps the type variable in another type is then reported as a nullability mismatch.

A named class extending the very same type with the very same overrides is accepted, and so is a local class declared inside a method. Only the anonymous form is affected.

### Reproducer

```java
package foo;

import java.util.List;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class Test {
  abstract static class Base<T extends @Nullable Object> {
    abstract List<T> make();
  }

  static Object anonymous() {
    return new Base<@Nullable String>() {
      // BUG: Diagnostic contains: Method returns List<@Nullable String>, but overridden method returns List<String>
      @Override List<@Nullable String> make() { throw new RuntimeException(); }
    };
  }

  // same supertype, same override: clean
  static class Named extends Base<@Nullable String> {
    @Override List<@Nullable String> make() { throw new RuntimeException(); }
  }
}
```

```
Test.java:16: warning: [NullAway] Method returns List<@Nullable String>, but overridden method returns List<String>, which has mismatched type parameter nullability
      @Override List<@Nullable String> make() { throw new RuntimeException(); }
                                       ^
```

### What does and does not report

| Declaration form | Reported |
| --- | --- |
| anonymous class extending `Base<@Nullable String>` | **yes** |
| anonymous class implementing `Iface<@Nullable String>` | **yes** |
| anonymous class in diamond form, `new Base<>() { … }` | **yes** |
| named nested class extending the same type | no |
| local class declared inside a method | no |
| anonymous class with a non-null type argument, `new Base<String>() { … }` | no |

| Override signature, inside the anonymous class | Reported |
| --- | --- |
| `@Nullable String get()` — bare type variable | no |
| `List<@Nullable String> get()` | **yes** |
| `List<List<@Nullable String>> get()` | **yes** |
| `@Nullable String[] get()` — array of the type variable | **yes** |
| `void take(List<@Nullable String> arg)` — parameter position | **yes** |
| an override that does not mention the type variable | no |

So the substitution is right at the top level of the override's type and wrong as soon as the type variable sits inside a type constructor. Both messages NullAway has for override mismatches appear:

```
Method returns List<@Nullable String>, but overridden method returns List<String>, which has mismatched type parameter nullability
Parameter has type List<@Nullable String>, but overridden method has parameter type List<String>, which has mismatched type parameter nullability
```

The local-class row is what separates this from a scoping problem: a class declared in the same statement position, with the same supertype, is checked correctly as soon as it has a name.

Neither the shape of the type argument nor the bound on the type variable changes the outcome; `@Nullable Object[]` as the type argument, and `T` declared with an implicit non-null bound, both behave the same as the case above. That is what separates it from #1732, where where the signature came from is what mattered.

### Why it is a false positive

`new Base<@Nullable String>() { … }` instantiates `Base` at `@Nullable String`, exactly as `class Named extends Base<@Nullable String>` does. The override's `List<@Nullable String>` is the substituted type of `List<T>`, so it matches. The diagnostic asks the author to write a return type that does not match the supertype.

### Workaround

Give the anonymous class a name. That keeps the checking in place, which `@SuppressWarnings("NullAway")` on the enclosing method would not.

### Where we hit it

`RedisTable.scan` returns an `Enumerable<@Nullable Object[]>` built as an anonymous `AbstractEnumerable<@Nullable Object[]>`. The same shape written as a named inner class, which is how `KafkaStreamTable` happens to spell it, is accepted. Found while replacing the Checker Framework with NullAway in Apache Calcite (https://github.com/apache/calcite/pull/5213).

**Comment by msridhar** [comment]
@vlsi it looks like this passes with the latest master branch; can you check?  I think #1722 may have fixed it

**Comment by vlsi** [comment]
Thanks. #1722 indeed resolves the current issue.

Confirmed by bisect: the reproducer fails at 70da029a and passes at a050aec4. Every shape from the table above is clean now, and `overrideExplicitlyTypedAnonymousClassWithNestedGenericTypes` already pins the behavior, so I see nothing missing for this issue.

Two follow-up tests I could send, if useful:

1. `new Base<>() { ... }` and `@Nullable String[]` for the anonymous-class case. Both work today, and both take paths the existing test does not reach: `inferCallType` for the diamond, `visitArrayType` for the array.
2. A call-site assertion for the astubx signature normalization. Dropping `replaceAll("\\s", "")` from `nullableReturns()` makes 36 signatures in `jspecify-jdk.astubx` match for the first time, all of them carrying wildcard bounds such as `? super K`. Those feed `hasNullableReturn` at every call site, not only at override checks. `mapComputeOverrideUsesJSpecifyModel` discards the result of `map.compute(...)`, so the call-site direction is unpinned.

That same change also surfaced what looks like a bad model entry. On master, `opt.orElseGet(() -> "x").length()` on an `Optional<String>` in `@NullMarked` code reports `dereferenced expression 'opt.orElseGet(() -> "x")' is @Nullable`; at 70da029a it reports nothing. `Optional:orElseGet(java.util.function.Supplier<? extends T>)` carries a `@Nullable` return in `jspecify-jdk.astubx`, and the old whitespace stripping was hiding it. The other 35 entries look correct: `Map.compute`, `merge`, and the `ConcurrentHashMap.search` family can all genuinely return null. Want me to open a separate issue for that one?

**Comment by msridhar** [comment]
Sending follow-up tests is fine!

Re: `Optional.orElseGet`, that signature with a `@Nullable` return is intentional, since the function can return `null` if the `Supplier` returns `null`.  I agree it will lead to a lot of noise, though.  To fix we need some kind of `@PolyNull` support, which is #1616.  In the meantime, if it adds too much noise, one can use the `IgnoreLibraryModelsFor` option ([docs](https://github.com/uber/NullAway/wiki/Configuration#library-models)).
